rclpy/rclpy/utilities.py

```diff
@@ -60,9 +60,19 @@ def shutdown(*, context=None):
 
 def try_shutdown(*, context=None):
     """Shutdown rclpy if not already shutdown."""
+    global g_context_lock
+    global g_default_context
     if context is None:
-        context = get_default_context()
-    return context.try_shutdown()
+        # Replace the default context with a new one if shutdown was successful
+        # or if the context was already shutdown.
+        with g_context_lock:
+            if g_default_context is None:
+                g_default_context = Context()
+            g_default_context.try_shutdown()
+            if not g_default_context.ok():
+                g_default_context = None
```

Comment:
I wish the default context registered an on_shutdown callback to manage the global variable too.

Reply:
Currently all the work is done with the default context lock taken.
If we register an on_shutdown() callback I don't think that would work.

Maybe it's possible to do that with some extra trick.